[DXIL] Add -fhlsl-unused-resource-bindings=reserve-all to ensure consistent binding assignments

Nielsbishere opened the pull request:
Adds a new compiler option: `-fhlsl-unused-resource-bindings=[strip|reserve-all]` .
`strip` keeps the current behavior.
`reserve-all` delays removing unused resources until after binding assignment, ensuring consistent register mappings across all entry points and library compilations when resources do not have explicit register annotations.

Under `reserve-all`, DCE removes only symbols initially, bindings are assigned with the complete resource list, and unused resources are stripped at the end. This produces deterministic and stable bindings without affecting optimization correctness.

Fixes https://github.com/microsoft/DirectXShaderCompiler/issues/7931

Nielsbishere committed:
Added rewriter option ConsistentBindings (-consistent-bindings) for the IDxcRewriter, this will allow generating register ids for registers that don't fully define them.

Nielsbishere committed:
Option -consistent-bindings now properly generates bindings for not fully qualified registers.

Nielsbishere committed:
Added a test for generating consistent bindings and changed tests to allow passing different rewriter flags. Also added support for cbuffer auto bindings.

Nielsbishere committed:
Applied clang format

Nielsbishere committed:
Now supporting multi dimensional register arrays for -consistent-bindings

Nielsbishere committed:
Removed rewriter option -consistent-bindings and now doing it in the dxil lowering properly, it's similar to -flegacy-resource-reservation but distinct in some key ways: the option only works for reserved registers and we need it to work on any register that has been compiled out. We also don't necessarily need the register to stay at the end of DXIL generation, in fact we'd rather not. Because having the register present can cause unintended performance penalties (for example, I use this reflection to know what transitions to issue)

Nielsbishere committed:
Reset some incorrect merges

Nielsbishere committed:
Cleanup for PR

Nielsbishere committed:
clang format

Nielsbishere committed:
Clang format again

Nielsbishere committed:
Added two unit tests; one for more complex scenarios (e.g. overlapping registers, $Globals, samplers, readonly, readwrite, etc.)

hekota commented:
@Nielsbishere - Looking at this change and your proposed update to add [-keep-all-resources](https://github.com/microsoft/DirectXShaderCompiler/pull/7734) flag, could you please help me understand what the benefit of the `-consistent-bindings` flag is compared to `-keep-all-resources`?

It seems to me that `-keep-all-resources` should already handle the consistent binding issue since it basically means “don’t strip unused resources.” The bindings in the example above end up inconsistent across different entry functions because the unused resources get stripped.

Also, as you’ve seen, implicit resource bindings tend to cause trouble since they make shader–application interfaces unpredictable and potentially compiler-dependent. It’d be great if you could find a way to stick with the explicit bindings to make sure your resource layouts stay stable and consistent.

tex3d commented on lib/HLSL/DxilCondenseResources.cpp:
```diff
@@ -571,6 +572,9 @@ class DxilLowerCreateHandleForLib : public ModulePass {
     bChanged |= ResourceRegisterAllocator.AllocateRegisters(DM);
 
+    if (DM.GetConsistentBindings())
+      DM.RemoveResourcesWithUnusedSymbols();
```
We probably need a test where `LegalizeResources` eliminates a resource use to test that code path. This relies on `DxilValueCache`, which was added largely to handle known-constant values in `-Od` cases where these values aren't simplified in the IR, but certain transformations we need to perform require these known values. In other words, you may be able to hit this code path using `-Od` with something that uses one resource or another in two code paths, depending on some condition that can be simplified to a constant, but is not normally simplified for `-Od`.

Nielsbishere replied:
Good catches, have adjusted RemoveResourcesWithUnusedSymbols to return isModified and passing it along to bChanged. Also fixed the issue you said with LegalizeResources. 
Since the issue ApplyBindingTableFromMetadata was already in there, it might be worth tackling this in a further PR/issue.
As for the test case for LegalizeResources; I'm not entirely sure where this would happen, I tried to read the code of LegalizeResources but it wasn't completely clear to me. Is there a unit test that tests legalize resources that I can combine with -consistent-bindings as a test case?

Nielsbishere commented:
@hekota excuse me for the (following) long message, but these two have two very different uses in my previous work place. I'm not maintaining there anymore, but I still want them to have these changes to continue my work and will tell them how it can be used.
The compile process is quite complex, so I won't fully go over it here, but essentially:
- Legacy technique syntax is parsed and replaced (PSO combinations, entrypoint definitions); coming from a DX9 codebase to DXR/DX12.
- Shader parsing & preprocessing is ran to acquire register bindings & even optimized out registers (and to ensure consistent bindings manually by inserting register annotations)
- Compilation & linking (shader variants with IDxcLinker)
- Shader reflection is used on the final binary to find non optimized out resources (for transitions and knowing if certain debug commands are used)

The pipeline going forward could be:
- Legacy technique syntax the same way, or a different system managing this in C++
- Compile lib with -keep-all-resources to retrieve all resources
- Compile & link each unique define combination with -consistent-bindings (getting rid of optimized out resources, but having consistent bindings)
- Shader reflection (same thing as before) and validate that no new registers are created or have mismatching bindings for safety (though never happens in practice)

# -keep-all-resources
is intended to (currently not working as it should in reflection, due to a missing createHandle) allow behavior similar to -Od with spirv; keep all resources even if optimized out. The reason why you don't want this by default is because reflection (presence of a register) maybe be used to guide for example transitions (or to know which real registers to keep). Why would you even want optimized out registers? Because sometimes you want a consistent interface that knows about the inputs to a shader, even if those shaders don't use the resource. It can still optimize it in the background but the interface will still show this variable as present. An example could be a visual scripting language that relies on this. The one I'm talking about relies on manual shader parsing to find out optimized out registers for the stability, adding this feature would allow to remove the manual parsing eventually. Consistent bindings in comparison, doesn't affect the emitted registers, only the bindings. So -keep-all-resources could be used to discover reflection info first and then a final compile per entrypoint could be used with -consistent-bindings to ensure no registers will be leftover.

# -consistent-bindings
is a separate issue in the same system. In modern or legacy pipelines, shaders might have been written without explicit bindings or legacy semantics. These get registers auto assigned. Once again, the same shader parser is used to replace these legacy semantics with register bindings to avoid this issue this PR fixes. A legit use case for this is poor man's bindless, where you have arrays of textures/buffers. You could have 1 central include that defines this layout and puts them at space1 (no explicit bind point). If you change the first array size or add a register to this include it will automatically update. Otherwise you need to adjust all subsequent resource bindings. The system I'm talking about also validates that the final shader reflection matches the expected bindings to avoid accidental mismatches.
With both of those solved, the parser can drop all HLSL preprocessing and rely on C++ annotations for the remainder that can't easily be parsed. I'm not maintaining this system anymore, but in my own framework this would also be useful.

Nielsbishere committed:
RemoveResourcesWithUnusedSymbolsHelper and RemoveResourcesWithUnusedSymbols now return bool isModified. Passing isModified along from RemoveResourcesWithUnusedSymbols to bChanged and fixed the LegalizeResources + -consistent-bindings oversight.

Nielsbishere committed:
Applied clang format

Nielsbishere committed:
Merge upstream into fork

Nielsbishere committed:
Fixed an issue where resources without a global variable would not properly trigger a modification

Nielsbishere committed:
Added -fhlsl-unused-resource-bindings<value> with options strip or reserve in this branch. Added release note

Nielsbishere committed:
Updated unit tests

Nielsbishere committed:
Merge branch 'main' of https://github.com/Microsoft/DirectXShaderCompiler into rewriter_generate_consistent_bindings

Nielsbishere committed:
Clang format that somehow doesn't get formatted correctly?

hekota commented on docs/ReleaseNotes.md:
```diff
@@ -23,6 +23,7 @@ The included licenses apply to the following files:
 - Fixed regression: [#7508](https://github.com/microsoft/DirectXShaderCompiler/issues/7508) crash when calling `Load` with `status`.
 - Header file `dxcpix.h` was added to the release package.
 - Moved Linear Algebra (Cooperative Vector) DXIL Opcodes to experimental Shader Model 6.10
+- Added `-fhlsl-unused-resource-bindings=<value>` an option to allow deciding on how to treat unused resource bindings in DXIL; `strip` (default), `keep` or `reserve`. `strip` will strip unused resources before generating bindings for resources without a `: register`, `keep` keeps them around while marking them with `D3D_SIF_UNUSED` in reflection and `reserve` will keep them reserved for generate ...
```
```suggestion
- Added `-fhlsl-unused-resource-bindings=<value>` an option to allow deciding on how to treat unused
  resource bindings in DXIL; `strip` (default), `keep-all` or `reserve-all`. `strip` will strip all unused resources,
  `keep-all` will keep them around while marking them with `D3D_SIF_UNUSED` in reflection, and `reserve` will keep
  their binding slots reserved when implicit bindings are assigned but not include them in the shader binding table
  or reflection. See [explanation] https://github.com/microsoft/DirectXShaderCompiler/pull/7643#issuecomment-3496917202)
  for more details.```

hekota replied:
Actually, just add the text for `reserve-all` for now. The `keep-all` should be added in the follow-up change.

hekota commented on include/dxc/Support/HLSLOptions.td:
```diff
@@ -576,6 +576,9 @@ def getprivate : JoinedOrSeparate<["-", "/"], "getprivate">, Flags<[DriverOption
   HelpText<"Suppress copyright message">;
 
+def fhlsl_unused_resource_bindings_EQ : Joined<["-"], "fhlsl-unused-resource-bindings=">, Group<hlslcomp_Group>, Flags<[CoreOption]>,
+  HelpText<"Control handling of unused resource bindings: 'strip' (default, generate bindings after optimization) or 'reserve'.">;
```
```suggestion
  HelpText<"Control handling of unused resource bindings: 'strip' (default, unused resources are stripped and theirs binding slots can be assigned to resources without explicit binding) or 'reserve-all' (do not use binding slots of unused resources when assigning bindings).">;
```

hekota commented on lib/HLSL/DxilCondenseResources.cpp:
```diff
@@ -550,27 +550,31 @@ class DxilLowerCreateHandleForLib : public ModulePass {
     unsigned newResources = DM.GetCBuffers().size() + DM.GetUAVs().size() +
                             DM.GetSRVs().size() + DM.GetSamplers().size();
-    bChanged = bChanged || (numResources != newResources);
+    bChanged |= numResources != newResources;
```
```suggestion
```
No longer needed because `bChanged` was already updated with the return value of `RemoveResourcesWithUnusedSymbols`?

hekota reviewed (commented):
Thank you for your contribution! As I mentioned above, we would like the options to 
be called `reserve-all` and `keep-all` to enable adding additional options like `reserve-explicit` in the future. 

If would be best if you could just add `reserve-all` value in this change, and once it is completed, you can update your other PR to add `keep-all` value to the enumeration. We generally want to avoid adding options that are not implemented or tested.

Also, the HLModule, DxilModule and CodeGenOptions should all use the UnusedResourceBindings enumeration instead of having separate flags for values (to avoid dealing with unsupported combinations of flags).

Nielsbishere committed:
PR feedback, re. reserve-all, bChanged and revert formatting done to release notes done by my markdown editor. TODO: Use UnusedResourceBindings in HLModule, DxilModule and CodeGenOptions

Nielsbishere committed:
Merge branch 'rewriter_generate_consistent_bindings' of https://github.com/Oxsomi/DirectXShaderCompiler into rewriter_generate_consistent_bindings

Nielsbishere committed:
Removed ConsistentBindings from intermediate flags and moved it to DxilShaderFlags instead. This is because 'keep-all' from the other PR relies on this Enum to be serialized; because reflection needs to know if unused resources are preserved.

Nielsbishere commented on docs/ReleaseNotes.md:
```diff
@@ -182,11 +183,11 @@ DX Compiler release for December 2022.
-experience.
+  the argument using the format `-ftime-trace=<filename>`. Chrome trace files can be opened in Chrome by loading the built-in tracing tool
+  at chrome://tracing. The trace file captures hierarchial timing data with additional context enabling a much more in-depth profiling
+  experience.
```
Must've been an accidental format by my markdown editor.

Nielsbishere commented:
No problem, fixed it. I moved the enum to DxilConstants.h, not sure if that's the right place but seems like it's the easiest to access from all of these various places. 
Instead of using Intermediate data I ended up adding it to the shader flags to ensure it lands into the final binary; this is required for the other PR since it needs to insert some shader variables to access the unbound resources in ID3D12FunctionReflection.

Nielsbishere committed:
Clang format

hekota commented on include/dxc/DXIL/DxilConstants.h:
```diff
@@ -15,6 +15,12 @@
+static_assert(UnusedResourceBinding::Count <= UnusedResourceBinding(7),
+              "Only 3 bits are reserved for UnusedResourceBinding by HLOptions "
+              "and ShaderFlags");
+
```
Love this!

hekota reviewed (commented):
Looking good and getting close! Could you please update the title and description to reflect the latest changes since that will become the commit message?

@tex3d - could you please comment on whether DxilConstants.h is a good place for the new enumeration?

tex3d commented on include/dxc/DXIL/DxilShaderFlags.h:
```diff
@@ -359,7 +368,9 @@ class ShaderFlags {
                                  // (OptFeatureInfo_RequiresGroup)
 
-  uint32_t m_align1 : 23; // align to 64 bit.
+  unsigned m_UnusedResourceBinding : 3;
```
I don't think we should be adding this setting into the `ShaderFlags`. These are visible to the driver and part of the DIXL contract, so we must be careful when adding flags here.

Instead, we need something like a LLVM module flag, which is stripped from the module when stripping reflection. It should be serialized along with HLModule and DxilModule metadata.

We already have metadata serialization for `LegacyResourceReservation`, through HLOptions serialization for HLModule and `IntermediateOptions` serialization in `DxilModule`. This is the right place for this option as well. The one issue is preservation of a flag indicating preservation of  `IntermediateOptions`

tex3d commented on lib/DXIL/DxilModule.cpp:
```diff
@@ -1021,8 +1030,9 @@ void DxilModule::RemoveUnusedResources() {
-static void RemoveResourcesWithUnusedSymbolsHelper(
+static bool RemoveResourcesWithUnusedSymbolsHelper(
     std::vector<std::unique_ptr<TResource>> &vec) {
+  bool modif = false;
```
The name `modif` has no precedent in this code base, we should use `Changed` instead. Elsewhere we might have used `bChanged`, but we have decided to try to follow our latest coding standards for new code wherever we can, even if it's in an existing function like this, as there isn't truly a single simple coding standard being followed by the existing code.
```suggestion
  bool Changed = false;
```

tex3d commented on lib/DXIL/DxilModule.cpp:
```diff
@@ -1044,14 +1055,17 @@ static void RemoveResourcesWithUnusedSymbolsHelper(
-  RemoveResourcesWithUnusedSymbolsHelper(m_CBuffers);
-  RemoveResourcesWithUnusedSymbolsHelper(m_Samplers);
+bool DxilModule::RemoveResourcesWithUnusedSymbols() {
+  bool modif = false;
```
Same suggestion here:
```suggestion
  bool Changed = false;
```

tex3d commented on include/dxc/DXIL/DxilConstants.h:
```diff
@@ -15,6 +15,12 @@
+
+static_assert(UnusedResourceBinding::Count <= UnusedResourceBinding(7),
+              "Only 3 bits are reserved for UnusedResourceBinding by HLOptions "
+              "and ShaderFlags");
```
Will need to update this comment when removing from ShaderFlags.

tex3d commented on lib/HLSL/HLModule.cpp:
```diff
@@ -231,17 +231,20 @@ void HLModule::RemoveFunction(llvm::Function *F) {
 
-    if (keepAllocated && (*p)->IsAllocated()) {
+    if ((keepAllocated && (*p)->IsAllocated()) ||
+        unusedResourceBinding == UnusedResourceBinding::ReserveAll) {
```
I think we could just unconditionally set the resource symbol to undef without removing it from the list, then rely on the `DxilCondenseResources` pass to clean up unused resources later as necessary. This simplifies reasoning about what code impacts this resource list too.

tex3d replied:
I think the logic looks correct. What sort of test failure are you seeing?

tex3d reviewed (changes requested):
I have requested several key changes:
- Don't use ShaderFlags for this, use HLSLOptions/IntermediateOptions instead.
- Don't bother removing from resource list in `HLModule::RemoveGlobal`, unconditionally set undef symbol instead.
- Don't condition calls to `RemoveResourcesWithUnusedSymbols` in `DxilCondenseResources`. Instead:
	- add `bool AfterAllocation=false` param
	- check `AfterAllocation`, `UnusedResourceBinding`, and `LegacyResourceReservation` (which will be folded into `UnusedResourceBinding`) for whether to remove resource
	- if not removing resource, set symbol to undef
- Coding standards / naming
- Need IR-based pass tests

NielsbishereAlt committed:
Changed modif to Changed

Nielsbishere committed:
Removed leftover b in front of bUnusedResourceBinding, removed unused resource bindings from shader flags. Added to IntermediateFlags instead. Updated static assert and enum cast.

Nielsbishere committed:
RemoveResourcesWithUnusedSymbolsHelper is now the one properly skipping removal of registers for ReserveAll it only kicks in after allocate registers.

Nielsbishere committed:
Removed leftover code

Nielsbishere committed:
Clang format and remove leftover dxil shader flag enum class

Nielsbishere commented on lib/HLSL/DxilCondenseResources.cpp:
Where do I add that? Any examples? (Same with the comment on DxilGenerationPass.cpp)

tex3d commented on include/dxc/DXIL/DxilModule.h:
```diff
@@ -345,7 +349,12 @@ class DxilModule {
+
     LegacyResourceReservation = 1 << 0,
+    UnusedResourceBindingMask =
+        7 << 1, // 3 bit reserved for unused resource binding
```
I think it would be easier to fold `LegacyResourceReservation` into `UnusedResourceBinding::ReserveExplicit`? That way you can simply serialize that state and get rid of `IntermediateFlags`.

Nielsbishere committed:
Moved legacyResourceReservation into UnusedResourceBinding::ReserveExplicit. Changed IntermediateFlags to m_UnusedResourceBindings. ClearIntermediateOptions->ResetUnusedResourceBinding. Left -flegacy-resource-reservation as is for now, though it routes to UnusedResourceBinding::ReserveExplicit internally.

Nielsbishere committed:
Clang format

Nielsbishere committed:
HLModule::RemoveGlobal now relies on DxilCondenseResources to remove resources. The problem before was that it never called RemoveResourcesWithUnusedSymbols on early exit. Also my earlier attempt tried to access *p while it should've accessed *c, since *p was already incremented.

Nielsbishere committed:
Clang format

damyanp commented on docs/ReleaseNotes.md:
```diff
@@ -23,6 +23,7 @@ The included licenses apply to the following files:
 - Fixed regression: [#7508](https://github.com/microsoft/DirectXShaderCompiler/issues/7508) crash when calling `Load` with `status`.
 - Header file `dxcpix.h` was added to the release package.
 - Moved Linear Algebra (Cooperative Vector) DXIL Opcodes to experimental Shader Model 6.10
+- Added `-fhlsl-unused-resource-bindings=<value>` an option to allow deciding on how to treat unused resource bindings in DXIL; `strip` (default) or `reserve-all`. `strip` will strip unused resources before generating bindings for resources without a `: register`, and `reserve-all` will keep them reserved for generated bindings while stripping them afterwards. See [explanation](https://github.com ...
```
Note for when resolving conflicts: make sure this is in the "Upcoming Release" section.  Automerge may well try and put this somewhere else.
